WMTS Get Cap document with updated WMTS.optionsFromCapabilities function

sarametz opened the pull request:
Building on pull request #3026 which fixes #2721, I altered the ol.source.WMTS.optionsFromCapabilities and ol.tilegrid.WMTS.createFromCapabilitiesMatrixSet functions to use the updated variable names (changed to be consistent with the WMS GetCapabilities format). This allows for WMTS sources to be easily created from a capabilities document. I have added a few tests for these and an example.

I have also altered the REST url template replacement to be case insensitive as some WMTS services use  {Style} and some use {style}. The WMTS spec seems to use the two interchangeably.

elemoine commented:
https://github.com/openlayers/ol3/pull/3026 is closed. This PR needs rebasing.

sarametz commented:
@elemoine I've rebased this PR, is there anything else I need to do before review?

probins commented on examples/wmts-layer-from-capabilities.js:
```diff
@@ -0,0 +1,32 @@
+    layers: [
+      new ol.layer.Tile({
+        opacity: 1,
+        extent: projectionExtent,
```
shouldn't the layer extent be taken from the BoundingBox in the capabilities doc if present?

elemoine commented on examples/wmts-layer-from-capabilities.js:
```diff
@@ -0,0 +1,32 @@
+$.ajax('data/WMTSCapabilities.xml').then(function(response) {
+  var result = parser.read(response);
+  var options = ol.source.WMTS.optionsFromCapabilities(result,
+      { layer: 'dem:oceanbottom', matrixSet: 'EPSG:3857'});
```
Minor, but please remove the space before `layer` here.

elemoine commented:
@sarametz, this looks great! Do we have a CLA from you already?

sarametz commented:
@probins using the bounding box from the capabilities document in that situation would require another function to create a tile layer from a capabilities document. So far in this PR I have just modified existing functions that were not currently working. @elemoine do you think it would be worth adding such a function? Also, I have removed the space.

sarametz commented:
@elemoine Koordinates Ltd. has submitted a CLA which I am included in

elemoine commented:
I just tested the example and I got an empty map. It looks like the WMTS you rely on in the example is down.

sarametz commented:
@elemoine I have updated the example to use a different WMTS layer, this involved replacing the capabilities document at examples/data/WMTSCapabilities.xml, which is only seems to be used in the example in #3026 which just parses the document

elemoine commented:
Thank you. I'll look at it once more and merge.

elemoine commented on src/ol/source/wmtssource.js:
```diff
@@ -239,90 +241,144 @@ ol.source.WMTS.prototype.updateDimensions = function(dimensions) {
+  var urls = [];
+  var requestEncoding = config['requestEncoding'];
+  if (wmtsCap['OperationsMetadata'].hasOwnProperty('GetTile') &&
+      requestEncoding != 'REST') {
```
We used to support both `REST` and `RESTful` as the request encoding string? Any reason for removing `RESTful`? I think it was there for a good reason.

sarametz replied:
@elemoine I didn't realise I had removed it, #3295 should add it back in

elemoine commented:
I just added a last comment regarding `RESTful` as the request encoding string. But, I'm going to merge this PR now anyway, as I am sure you will come with a new PR if my comment makes sense :)

Thanks a lot for your good patches @sarametz!
